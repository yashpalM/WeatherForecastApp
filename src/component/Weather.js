import React from 'react';
import { ActivityIndicator, FlatList, StyleSheet, Text, View } from 'react-native';
import { Card } from 'react-native-elements';
import Icon from 'react-native-vector-icons/Ionicons';
import colors from '../constants/colors'
import utils from '../utils/Utils';
import _ from 'lodash';

const generateIcon = (description, sys) => {
const { pod } = sys;
const dayTime = pod === 'd';
let icon;
switch (description) {
    case 'Haze':
        icon = dayTime ?
            (<Icon name="ios-cloudy" style={styles.weatherIcon} size={40} color="#FFFFFF" />) :
            (<Icon name="ios-cloudy-night" style={styles.weatherIcon} size={40} color="#FFFFFF" />);
        break;
    case 'Snow':
        icon = (<Icon name="ios-snow" style={styles.weatherIcon} size={40} color="#FFFFFF" />);
        break;
    case 'Clouds':
        icon = (<Icon name="ios-cloudy" style={styles.weatherIcon} size={40} color="#FFFFFF" />);
        break;
    case 'Clear':
        icon = dayTime ?
            (<Icon name="ios-sunny" style={styles.weatherIcon} size={40} color="#FFFF44" />) :
            (<Icon name="ios-moon" style={styles.weatherIcon} size={40} color="#FFFFFF" />);
        break;
    case 'Rain':
        icon = (<Icon name="ios-rainy" style={styles.weatherIcon} size={40} color="#FFFFFF" />);
        break;
    case 'Fog':
        icon = (<Icon name="ios-partly-sunny" style={styles.weatherIcon} size={40} color="#FFFFFF" />);
        break;
    default:
        icon = (<Icon name="ios-sunny" style={styles.weatherIcon} size={40} color="#FFFFFF" />);
        break;
}
return icon;
};


const renderWeatherContent = data => (
<View>
    {_.isEmpty(data) ?
        <Text style={styles.emptyData} >{'Search for a location'}</Text>
        :
        <View style={{ marginBottom: 12 }}>

            <Text style={styles.title}>{data.city.name},{data.city.country}</Text>

            <FlatList
                data={data.list}
                contentContainerStyle={{ paddingBottom: 200 }}
                keyExtractor={({ dt }, index) => dt.toString()}
                renderItem={({ item }) =>
                    <Card containerStyle={styles.cardContainer}>
                        <View style={styles.viewContainer}>
                            <View style={{ flex: 1, alignSelf: 'center' }}>
                                <Text style={styles.description}>{`${utils.dateToFormat(item.dt_txt)}`}</Text>
                            </View>

                            <View style={styles.tempContainer}>
                                <Text style={styles.tempratureText}>{`${utils.toCelsius(item.main.temp)} Â°C`}</Text>
                                <Text style={styles.weatherDesc}>{item.weather[0].main},{item.weather[0].description}</Text>
                            </View>

                            <View style={{ flex: 1, alignSelf: 'center' }}>
                                {generateIcon(item.weather[0].main, item.sys)}
                            </View>

                        </View>

                    </Card>
                }
            />
        </View>

    }
</View>
)

const errorRendering = errorMessage => (
<View style={styles.errorContainer}>
    <Text style={styles.errorText}>{errorMessage}</Text>
</View>
)


const Weather = props => {
const { data, errorMessage, isLoading } = props
const container = _.isEmpty(errorMessage) ?
    renderWeatherContent(data) :
    errorRendering(errorMessage);
return (
    <View>
        {isLoading ?
            <ActivityIndicator color={colors.primary} size='large' /> :
            container
        }
    </View>
);
};

const styles = StyleSheet.create({
emptyData: {
    textAlign: 'center',
    fontSize: 20,
    margin: 10,
    color: colors.white,
},
title: {
    fontSize: 24,
    fontWeight: 'bold',
    color: colors.black,
    textAlign: 'center',
    marginBottom: 10,
    marginTop: 10,
},
cardContainer: {
    elevation: 5,
    backgroundColor: colors.primary
},
cardContainer: {
    elevation: 5,
    backgroundColor: colors.primary
},
viewContainer: {
    flex: 1,
    width: 280,
    alignItems: 'flex-start',
    padding: 5,
    flexDirection: 'row',
},
description: {
    fontSize: 16,
    fontWeight: '400',
    color: colors.white,
    textAlign: 'center',
},
tempContainer: {
    flex: 2,
    marginHorizontal: 20,
    flexGrow: 2,
    flexDirection: 'column'
},
tempratureText: {
    color: '#f4511e',
    textAlignVertical: 'center',
    textAlign: 'center',
    alignSelf: 'stretch',
    fontSize: 22,
    fontWeight: 'bold',
},
weatherDesc: {
    color: colors.white,
    textAlignVertical: 'center',
    textAlign: 'center',
    alignSelf: 'stretch'
},
errorContainer: {
    borderColor: colors.red,
    borderWidth: 1,
    backgroundColor: colors.pink,
    padding: 5,
},
errorText: {
    textAlign: 'center',
    color: colors.red,
    fontSize: 20,
    margin: 10,
},
weatherIcon: {
    textAlign: 'center',
    justifyContent: 'center',
    alignItems: 'center',
    textAlignVertical: 'center',
    textAlign: 'center',
},
});

export default Weather;