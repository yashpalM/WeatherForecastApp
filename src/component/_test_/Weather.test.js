import 'react-native';
import renderer from 'react-test-renderer';
import React from 'react';
import { shallow } from 'enzyme';
import Weather from '../Weather';


const mockWeatherData = require('../../../_mocks_/weatherResponse.json');

it('Weather component renders correctly', () => {
    renderer.create(<Weather />);
});

describe('Weather', () => {
    describe('Rendering', () => {
        it('should match to snapshot - Primary', () => {
            const component = shallow(<Weather label="test label" isLoading />)
            expect(component).toMatchSnapshot("Primary Weather snapshot")
        });

        it('should match to snapshot - Secondary', () => {
            const component = shallow(<Weather label="test label" isLoading={false} />)
            expect(component).toMatchSnapshot("Secondary Weather snapshot")
        });

        it('should match to snapshot', () => {
            const component = shallow(<Weather errorMessage='Could not fetch weatherData' />)
            expect(component).toMatchSnapshot()
        });

        it('should render correctly when loading', () => {
            const componet = shallow(<Weather label="test label" isLoading />)
            expect(componet.find('View').length).toBe(1);
        });

        it('should render correctly when loading false and errorMessage', () => {
            const component = shallow(<Weather label="test label" isLoading={false} errorMessage='Test' />)
            expect(component.find('View').length).toBe(2);
            expect(component.find('Text').length).toBe(1);
        });

        it('shhould render correctly when loading false and mock weatherData', () => {
            const component = shallow(<Weather isLoading={false} errorMessage='' data={mockWeatherData} />)
            expect(component.find('View').length).toBe(3);
            expect(component.find('Text').length).toBe(1);
            expect(component.find('FlatList').length).toBe(1);
        });

        it('should render correctly when loading false and empty weatherData', () => {
            const component = shallow(<Weather isLoading={false} errorMessage='' />)
            expect(component.find('View').length).toBe(2);
            expect(component.find('Text').length).toBe(1);
        });

        it('should flatlist return KeyExtractor correctly', () => {
            const component = shallow(<Weather isLoading={false} errorMessage='' data={mockWeatherData} />)
            const key = component.find('FlatList').props().keyExtractor({ dt: 1618077000 });
            expect(key).toEqual('1618077000')
        });

    });

    describe('.renderItem', () => {
        let renderItemShallowWrapper;
        const mockItem = mockWeatherData.list[0];
        let wrapper;
        let RenderItem;

        beforeAll(() => {
            wrapper = shallow(<Weather isLoading={false} errorMessage='' data={mockWeatherData} />)
            // find the component whose property is rendered as renderItem={[Function]}
            RenderItem = wrapper.find('FlatList').prop('renderItem');
        });

        it('should match the snapshot', () => {
            renderItemShallowWrapper = shallow(<RenderItem item={mockItem} />);
            expect(renderItemShallowWrapper).toMatchSnapshot();
        });

        it('should match the snapshot - Haze', () => {
            mockItem.weather[0].main = 'Haze';
            // and since it's a component render it as such
            // with mockItem
            renderItemShallowWrapper = shallow(<RenderItem item={mockItem} />);
            // generate snapshot for the renderItem
            expect(renderItemShallowWrapper).toMatchSnapshot();
        });

        it('should match the snapshot - Snow', () => {
            mockItem.weather[0].main = 'Snow';
            renderItemShallowWrapper = shallow(<RenderItem item={mockItem} />);
            expect(renderItemShallowWrapper).toMatchSnapshot();
        });

        it('should match the snapshot - Clear', () => {
            mockItem.weather[0].main = 'Clear';
            renderItemShallowWrapper = shallow(<RenderItem item={mockItem} />);
            expect(renderItemShallowWrapper).toMatchSnapshot();
        });

        it('should match the snapshot - Rain', () => {
            mockItem.weather[0].main = 'Rain';
            renderItemShallowWrapper = shallow(<RenderItem item={mockItem} />);
            expect(renderItemShallowWrapper).toMatchSnapshot();
        });

        it('should match the snapshot - Fog', () => {
            mockItem.weather[0].main = 'Fog';
            renderItemShallowWrapper = shallow(<RenderItem item={mockItem} />);
            expect(renderItemShallowWrapper).toMatchSnapshot();
        });

        it('should match the snapshot - Default', () => {
            mockItem.weather[0].main = 'abc';
            renderItemShallowWrapper = shallow(<RenderItem item={mockItem} />);
            expect(renderItemShallowWrapper).toMatchSnapshot();
        });

        it('should render correctly - renderItem - View', () => {
            expect(renderItemShallowWrapper.find('View').length).toBe(4);
        });

        it('should render correctly - renderItem - Text', () => {
            expect(renderItemShallowWrapper.find('Text').length).toBe(3);
        });

    });
});

