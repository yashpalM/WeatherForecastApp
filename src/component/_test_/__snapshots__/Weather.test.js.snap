// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather .renderItem should match the snapshot - Clear 1`] = `ShallowWrapper {}`;

exports[`Weather .renderItem should match the snapshot - Default 1`] = `ShallowWrapper {}`;

exports[`Weather .renderItem should match the snapshot - Fog 1`] = `ShallowWrapper {}`;

exports[`Weather .renderItem should match the snapshot - Haze 1`] = `ShallowWrapper {}`;

exports[`Weather .renderItem should match the snapshot - Rain 1`] = `ShallowWrapper {}`;

exports[`Weather .renderItem should match the snapshot - Snow 1`] = `ShallowWrapper {}`;

exports[`Weather .renderItem should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`Weather Rendering should match to snapshot - Primary: Primary Weather snapshot 1`] = `ShallowWrapper {}`;

exports[`Weather Rendering should match to snapshot - Secondary: Secondary Weather snapshot 1`] = `ShallowWrapper {}`;

exports[`Weather Rendering should match to snapshot 1`] = `ShallowWrapper {}`;
